File: code/go/ch07/src/luago/vm/opcodes.go
package vm

import "luago/api"

/* OpMode */
/* basic instruction format */
const (
	IABC  = iota // [  B:9  ][  C:9  ][ A:8  ][OP:6]
	IABx         // [      Bx:18     ][ A:8  ][OP:6]
	IAsBx        // [     sBx:18     ][ A:8  ][OP:6]
	IAx          // [           Ax:26        ][OP:6]
)

/* OpArgMask */
const (
	OpArgN = iota // argument is not used
	OpArgU        // argument is used
	OpArgR        // argument is a register or a jump offset
	OpArgK        // argument is a constant or register/constant
)

/* OpCode */
const (
	OP_MOVE = iota
	OP_LOADK
	OP_LOADKX
	OP_LOADBOOL
	OP_LOADNIL
	OP_GETUPVAL
	OP_GETTABUP
	OP_GETTABLE
	OP_SETTABUP
	OP_SETUPVAL
	OP_SETTABLE
	OP_NEWTABLE
	OP_SELF
	OP_ADD
	OP_SUB
	OP_MUL
	OP_MOD
	OP_POW
	OP_DIV
	OP_IDIV
	OP_BAND
	OP_BOR
	OP_BXOR
	OP_SHL
	OP_SHR
	OP_UNM
	OP_BNOT
	OP_NOT
	OP_LEN
	OP_CONCAT
	OP_JMP
	OP_EQ
	OP_LT
	OP_LE
	OP_TEST
	OP_TESTSET
	OP_CALL
	OP_TAILCALL
	OP_RETURN
	OP_FORLOOP
	OP_FORPREP
	OP_TFORCALL
	OP_TFORLOOP
	OP_SETLIST
	OP_CLOSURE
	OP_VARARG
	OP_EXTRAARG
)

type opcode struct {
	testFlag byte // operator is a test (next instruction must be a jump)
	setAFlag byte // instruction set register A
	argBMode byte // B arg mode
	argCMode byte // C arg mode
	opMode   byte // op mode
	name     string
	action   func(i Instruction, vm api.LuaVM)
}

var opcodes = []opcode{
	/*     T  A    B       C     mode         name       action */
	opcode{0, 1, OpArgR, OpArgN, IABC /* */, "MOVE    ", move},     // R(A) := R(B)
	opcode{0, 1, OpArgK, OpArgN, IABx /* */, "LOADK   ", loadK},    // R(A) := Kst(Bx)
	opcode{0, 1, OpArgN, OpArgN, IABx /* */, "LOADKX  ", loadKx},   // R(A) := Kst(extra arg)
	opcode{0, 1, OpArgU, OpArgU, IABC /* */, "LOADBOOL", loadBool}, // R(A) := (bool)B; if (C) pc++
	opcode{0, 1, OpArgU, OpArgN, IABC /* */, "LOADNIL ", loadNil},  // R(A), R(A+1), ..., R(A+B) := nil
	opcode{0, 1, OpArgU, OpArgN, IABC /* */, "GETUPVAL", nil},      // R(A) := UpValue[B]
	opcode{0, 1, OpArgU, OpArgK, IABC /* */, "GETTABUP", nil},      // R(A) := UpValue[B][RK(C)]
	opcode{0, 1, OpArgR, OpArgK, IABC /* */, "GETTABLE", nil},      // R(A) := R(B)[RK(C)]
	opcode{0, 0, OpArgK, OpArgK, IABC /* */, "SETTABUP", nil},      // UpValue[A][RK(B)] := RK(C)
	opcode{0, 0, OpArgU, OpArgN, IABC /* */, "SETUPVAL", nil},      // UpValue[B] := R(A)
	opcode{0, 0, OpArgK, OpArgK, IABC /* */, "SETTABLE", nil},      // R(A)[RK(B)] := RK(C)
	opcode{0, 1, OpArgU, OpArgU, IABC /* */, "NEWTABLE", nil},      // R(A) := {} (size = B,C)
	opcode{0, 1, OpArgR, OpArgK, IABC /* */, "SELF    ", nil},      // R(A+1) := R(B); R(A) := R(B)[RK(C)]
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "ADD     ", add},      // R(A) := RK(B) + RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "SUB     ", sub},      // R(A) := RK(B) - RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "MUL     ", mul},      // R(A) := RK(B) * RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "MOD     ", mod},      // R(A) := RK(B) % RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "POW     ", pow},      // R(A) := RK(B) ^ RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "DIV     ", div},      // R(A) := RK(B) / RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "IDIV    ", idiv},     // R(A) := RK(B) // RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "BAND    ", band},     // R(A) := RK(B) & RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "BOR     ", bor},      // R(A) := RK(B) | RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "BXOR    ", bxor},     // R(A) := RK(B) ~ RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "SHL     ", shl},      // R(A) := RK(B) << RK(C)
	opcode{0, 1, OpArgK, OpArgK, IABC /* */, "SHR     ", shr},      // R(A) := RK(B) >> RK(C)
	opcode{0, 1, OpArgR, OpArgN, IABC /* */, "UNM     ", unm},      // R(A) := -R(B)
	opcode{0, 1, OpArgR, OpArgN, IABC /* */, "BNOT    ", bnot},     // R(A) := ~R(B)
	opcode{0, 1, OpArgR, OpArgN, IABC /* */, "NOT     ", not},      // R(A) := not R(B)
	opcode{0, 1, OpArgR, OpArgN, IABC /* */, "LEN     ", length},   // R(A) := length of R(B)
	opcode{0, 1, OpArgR, OpArgR, IABC /* */, "CONCAT  ", concat},   // R(A) := R(B).. ... ..R(C)
	opcode{0, 0, OpArgR, OpArgN, IAsBx /**/, "JMP     ", jmp},      // pc+=sBx; if (A) close all upvalues >= R(A - 1)
	opcode{1, 0, OpArgK, OpArgK, IABC /* */, "EQ      ", eq},       // if ((RK(B) == RK(C)) ~= A) then pc++
	opcode{1, 0, OpArgK, OpArgK, IABC /* */, "LT      ", lt},       // if ((RK(B) <  RK(C)) ~= A) then pc++
	opcode{1, 0, OpArgK, OpArgK, IABC /* */, "LE      ", le},       // if ((RK(B) <= RK(C)) ~= A) then pc++
	opcode{1, 0, OpArgN, OpArgU, IABC /* */, "TEST    ", test},     // if not (R(A) <=> C) then pc++
	opcode{1, 1, OpArgR, OpArgU, IABC /* */, "TESTSET ", testSet},  // if (R(B) <=> C) then R(A) := R(B) else pc++
	opcode{0, 1, OpArgU, OpArgU, IABC /* */, "CALL    ", nil},      // R(A), ... ,R(A+C-2) := R(A)(R(A+1), ... ,R(A+B-1))
	opcode{0, 1, OpArgU, OpArgU, IABC /* */, "TAILCALL", nil},      // return R(A)(R(A+1), ... ,R(A+B-1))
	opcode{0, 0, OpArgU, OpArgN, IABC /* */, "RETURN  ", nil},      // return R(A), ... ,R(A+B-2)
	opcode{0, 1, OpArgR, OpArgN, IAsBx /**/, "FORLOOP ", forLoop},  // R(A)+=R(A+2); if R(A) <?= R(A+1) then { pc+=sBx; R(A+3)=R(A) }
	opcode{0, 1, OpArgR, OpArgN, IAsBx /**/, "FORPREP ", forPrep},  // R(A)-=R(A+2); pc+=sBx
	opcode{0, 0, OpArgN, OpArgU, IABC /* */, "TFORCALL", nil},      // R(A+3), ... ,R(A+2+C) := R(A)(R(A+1), R(A+2));
	opcode{0, 1, OpArgR, OpArgN, IAsBx /**/, "TFORLOOP", nil},      // if R(A+1) ~= nil then { R(A)=R(A+1); pc += sBx }
	opcode{0, 0, OpArgU, OpArgU, IABC /* */, "SETLIST ", nil},      // R(A)[(C-1)*FPF+i] := R(A+i), 1 <= i <= B
	opcode{0, 1, OpArgU, OpArgN, IABx /* */, "CLOSURE ", nil},      // R(A) := closure(KPROTO[Bx])
	opcode{0, 1, OpArgU, OpArgN, IABC /* */, "VARARG  ", nil},      // R(A), R(A+1), ..., R(A+B-2) = vararg
	opcode{0, 0, OpArgU, OpArgU, IAx /*  */, "EXTRAARG", nil},      // extra (larger) argument for previous opcode
}

File: code/csharp/ch09/luavm/vm/OpCodes.cs
using LuaVM = luavm.state.LuaState;

namespace luavm.vm
{
    internal struct opcode
    {
        internal byte testFlag;
        internal byte setAFlag;
        internal byte argBMode;
        internal byte argCMode;
        internal byte opMode;
        internal string name;
        internal Action action;

        public opcode(byte testFlag, byte aFlag, byte argBMode, byte argCMode, byte opMode, string name, Action action)
        {
            this.testFlag = testFlag;
            setAFlag = aFlag;
            this.argBMode = argBMode;
            this.argCMode = argCMode;
            this.opMode = opMode;
            this.name = name;
            this.action = action;
        }
    }

    delegate void Action(Instruction i, ref LuaVM vm);

    public static class OpCodes
    {
        private const byte IABC = 0;
        private const byte IABx = 1;
        private const byte IAsBx = 2;
        private const byte IAx = 3;

        internal const byte OP_MOVE = 0;
        internal const byte OP_LOADK = 1;
        internal const byte OP_LOADKX = 2;
        internal const byte OP_LOADBOOL = 3;
        internal const byte OP_LOADNIL = 4;
        internal const byte OP_GETUPVAL = 5;
        internal const byte OP_GETTABUP = 6;
        internal const byte OP_GETTABLE = 7;
        internal const byte OP_SETTABUP = 8;
        internal const byte OP_SETUPVAL = 9;
        internal const byte OP_SETTABLE = 10;
        internal const byte OP_NEWTABLE = 11;
        internal const byte OP_SELF = 12;
        internal const byte OP_ADD = 13;
        internal const byte OP_SUB = 14;
        internal const byte OP_MUL = 15;
        internal const byte OP_MOD = 16;
        internal const byte OP_POW = 17;
        internal const byte OP_DIV = 18;
        internal const byte OP_IDIV = 19;
        internal const byte OP_BAND = 20;
        internal const byte OP_BOR = 21;
        internal const byte OP_BXOR = 22;
        internal const byte OP_SHL = 23;
        internal const byte OP_SHR = 24;
        internal const byte OP_UNM = 25;
        internal const byte OP_BNOT = 26;
        internal const byte OP_NOT = 27;
        internal const byte OP_LEN = 28;
        internal const byte OP_CONCAT = 29;
        internal const byte OP_JMP = 30;
        internal const byte OP_EQ = 31;
        internal const byte OP_LT = 32;
        internal const byte OP_LE = 33;
        internal const byte OP_TEST = 34;
        internal const byte OP_TESTSET = 35;
        internal const byte OP_CALL = 36;
        internal const byte OP_TAILCALL = 37;
        internal const byte OP_RETURN = 38;
        internal const byte OP_FORLOOP = 39;
        internal const byte OP_FORPREP = 40;
        internal const byte OP_TFORCALL = 41;
        internal const byte OP_TFORLOOP = 42;
        internal const byte OP_SETLIST = 43;
        internal const byte OP_CLOSURE = 44;
        internal const byte OP_VARARG = 45;
        internal const byte OP_EXTRAARG = 46;

        internal const byte OpArgN = 0;
        internal const byte OpArgU = 1;
        internal const byte OpArgR = 2;
        internal const byte OpArgK = 3;

        internal static opcode[] opcodes =
        {
            /*   T            A             B                     C                   mode         name              action*/
            new opcode(0, 1, OpArgR, OpArgN, IABC, "MOVE    ", InstMisc.move), // R(A) := R(B)
            new opcode(0, 1, OpArgK, OpArgN, IABx, "LOADK   ", InstLoad.loadK), // R(A) := Kst(Bx)
            new opcode(0, 1, OpArgN, OpArgN, IABx, "LOADKX  ", InstLoad.loadKx), // R(A) := Kst(extra arg)
            new opcode(0, 1, OpArgU, OpArgU, IABC, "LOADBOOL", InstLoad.loadBool), // R(A) := (bool)B; if (C) pc++
            new opcode(0, 1, OpArgU, OpArgN, IABC, "LOADNIL ",
                InstLoad.loadNil), // R(A), R(A+1), ..., R(A+B) := nil
            new opcode(0, 1, OpArgU, OpArgN, IABC, "GETUPVAL", null), // R(A) := UpValue[B]
            new opcode(0, 1, OpArgU, OpArgK, IABC, "GETTABUP", InstUpvalue.getTabUp), // R(A) := UpValue[B][RK(C)]
            new opcode(0, 1, OpArgR, OpArgK, IABC /* */, "GETTABLE", InstTable.getTable), // R(A) := R(B)[RK(C)]
            new opcode(0, 0, OpArgK, OpArgK, IABC /* */, "SETTABUP", null), // UpValue[A][RK(B)] := RK(C)
            new opcode(0, 0, OpArgU, OpArgN, IABC /* */, "SETUPVAL", null), // UpValue[B] := R(A)
            new opcode(0, 0, OpArgK, OpArgK, IABC /* */, "SETTABLE", InstTable.setTable), // R(A)[RK(B)] := RK(C)
            new opcode(0, 1, OpArgU, OpArgU, IABC /* */, "NEWTABLE", InstTable.newTable), // R(A) := {} (size = B,C)
            new opcode(0, 1, OpArgR, OpArgK, IABC /* */, "SELF    ", null), // R(A+1) := R(B); R(A) := R(B)[RK(C)]
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "ADD     ", InstOperators.add), // R(A) := RK(B) + RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "SUB     ", InstOperators.sub), // R(A) := RK(B) - RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "MUL     ", InstOperators.mul), // R(A) := RK(B) * RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "MOD     ", InstOperators.mod), // R(A) := RK(B) % RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "POW     ", InstOperators.pow), // R(A) := RK(B) ^ RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "DIV     ", InstOperators.div), // R(A) := RK(B) / RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "IDIV    ", InstOperators.idiv), // R(A) := RK(B) // RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "BAND    ", InstOperators.band), // R(A) := RK(B) & RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "BOR     ", InstOperators.bor), // R(A) := RK(B) | RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "BXOR    ", InstOperators.bxor), // R(A) := RK(B) ~ RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "SHL     ", InstOperators.shl), // R(A) := RK(B) << RK(C)
            new opcode(0, 1, OpArgK, OpArgK, IABC /* */, "SHR     ", InstOperators.shr), // R(A) := RK(B) >> RK(C)
            new opcode(0, 1, OpArgR, OpArgN, IABC /* */, "UNM     ", InstOperators.unm), // R(A) := -R(B)
            new opcode(0, 1, OpArgR, OpArgN, IABC /* */, "BNOT    ", InstOperators.bnot), // R(A) := ~R(B)
            new opcode(0, 1, OpArgR, OpArgN, IABC /* */, "NOT     ", InstOperators.not), // R(A) := not R(B)
            new opcode(0, 1, OpArgR, OpArgN, IABC /* */, "LEN     ",
                InstOperators.length), // R(A) := length of R(B)
            new opcode(0, 1, OpArgR, OpArgR, IABC /* */, "CONCAT  ",
                InstOperators.concat), // R(A) := R(B).. ... ..R(C)
            new opcode(0, 0, OpArgR, OpArgN, IAsBx, "JMP     ", InstMisc.jmp),
            new opcode(1, 0, OpArgK, OpArgK, IABC, "EQ      ", InstOperators.eq),
            new opcode(1, 0, OpArgK, OpArgK, IABC, "LT      ", InstOperators.lt),
            new opcode(1, 0, OpArgK, OpArgK, IABC, "LE      ", InstOperators.le),
            new opcode(1, 0, OpArgN, OpArgU, IABC, "TEST    ", InstOperators.test),
            new opcode(1, 1, OpArgR, OpArgU, IABC, "TESTSET ", InstOperators.testSet),
            new opcode(0, 1, OpArgU, OpArgU, IABC /* */, "CALL    ", InstCall.call),
            new opcode(0, 1, OpArgU, OpArgU, IABC /* */, "TAILCALL",
                InstCall.tailCall), // return R(A)(R(A+1), ... ,R(A+B-1))
            new opcode(0, 0, OpArgU, OpArgN, IABC /* */, "RETURN  ",
                InstCall._return), // return R(A), ... ,R(A+B-2)
            new opcode(0, 1, OpArgR, OpArgN, IAsBx /**/, "FORLOOP ", InstFor.forLoop),
            new opcode(0, 1, OpArgR, OpArgN, IAsBx /**/, "FORPREP ", InstFor.forPrep), // R(A)-=R(A+2); pc+=sBx
            new opcode(0, 0, OpArgN, OpArgU, IABC /* */, "TFORCALL", null),
            new opcode(0, 1, OpArgR, OpArgN, IAsBx /**/, "TFORLOOP", null),
            new opcode(0, 0, OpArgU, OpArgU, IABC /* */, "SETLIST ", InstTable.setList),
            new opcode(0, 1, OpArgU, OpArgN, IABx /* */, "CLOSURE ", InstCall.closure),
            new opcode(0, 1, OpArgU, OpArgN, IABC /* */, "VARARG  ", InstCall.vararg),
            new opcode(0, 0, OpArgU, OpArgU, IAx /*  */, "EXTRAARG", null),
        };
    }
}